flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/graph/TestEventSource.java

```diff
@@ -103,6 +110,14 @@ private void ack(TestCommand cmd) {
 
     @Override
     public void cancel() {
+        stop();
+    }
+
+    private void stop() {
+        commandQueue.unsubscribe(operatorID, commandExecutor);
         isRunning = false;
+        if (!scheduledCommands.isEmpty()) {
```

Comment:
Should `scheduledCommands` be volatile as well? IIUC `open + run` and `stop` could be executed from different threads.

Reply:
Good point!
Fixed in 783cc00fae8138259209d755349c4ce296e478c0.